![image.png](attachment:image.png)

## PRACTICA OBLIGATORIA: **Álgebra y NumPy**

* La práctica obligatoria de esta unidad consiste en dos ejercicios de programación libre a completar. Descarga este notebook en tu ordenador y trabaja en local.  
* Recuerda que debes subirla a tu repositorio personal antes de la sesión en vivo para que puntúe adecuadamente.  
* Recuerda también que no es necesario que esté perfecta, sólo es necesario que se vea el esfuerzo. 
* Esta práctica se resolverá en la sesión en vivo correspondiente y la solución se publicará en el repo del curso. 

![image.png](attachment:image.png)

## PARTE I:

<img src ='https://estaticos-cdn.prensaiberica.es/clip/23bf0054-14a8-40d0-8700-dc653638d702_16-9-discover-aspect-ratio_default_0.jpg' width=300>

Rafa, Nacho y Clara quieren comprar palmeras, donuts, hojaldres y tartas. Cada uno necesita diferentes cantidades y pueden elegir dos tiendas: *Manolo Bakes* y *Starbucks*. 

Las siguiente tablas muestran, respectivamente, la cantidad que cada uno necesita, y los precios en cada una de las tiendas

<table><thead><tr><th></th><th>Palmeras</th><th>Donuts</th><th>Hojaldres</th><th>Tartas</th></tr></thead><tbody><tr><td>Rafa</td><td>6</td><td>5</td><td>3</td><td>1</td></tr><tr><td>Nacho</td><td>3</td><td>6</td><td>2</td><td>2</td></tr><tr><td>Clara</td><td>3</td><td>4</td><td>3</td><td>1</td></tr></tbody></table>
<br></br>
<table><thead><tr><th></th><th>Manolo Bakes</th><th>Starbucks</th></tr></thead><tbody><tr><td>Palmera</td><td>1.50</td><td>1.00</td></tr><tr><td>Donut</td><td>2.00</td><td>2.50</td></tr><tr><td>Hojaldre</td><td>5.00</td><td>4.50</td></tr><tr><td>Tarta</td><td>16.00</td><td>17.00</td></tr></tbody></table>

Por ejemplo, si Rafa compra en *Manolo Bakes* gastaría:  

$6\cdot1.50+5\cdot2+3\cdot5+1\cdot16=50$  

y si compra en *Starbucks* gastaría:  

$6\cdot1+5\cdot2.50+3\cdot4.50+1\cdot17=49$  

**SE PIDE**: Genera la matriz que exprese la cantidad que cada persona gastaría en cada tienda, y recomienda a cada uno dónde debería hacer sus compras (pista: donde le cueste menos, ¿no?)

**REQUISITIO**: Emplea Numpy, su tipo array y sus operaciones para la resolución del ejercicio. 

In [ ]:
import numpy as np

In [ ]:
mat_bollos = np.array([
    [6, 5, 3, 1],
    [3, 6, 2, 2],
    [3, 4, 3, 1]
])

print(mat_bollos)

[[6 5 3 1]
 [3 6 2 2]
 [3 4 3 1]]


In [ ]:
# Precios por unidad en cada cafetería
tarifas_manolo = np.array([1.5, 2.0, 5.0, 16.0])
tarifas_starbucks = np.array([1.0, 2.5, 4.5, 17.0])

# Matriz 2x4 con los precios de ambas cafeterías
mat_precios = np.array([tarifas_manolo, tarifas_starbucks])

print(mat_precios)



[[ 1.5  2.   5.  16. ]
 [ 1.   2.5  4.5 17. ]]


In [ ]:
# Transponemos para que las dimensiones cuadren en el producto matricial
mat_precios = mat_precios.T  # ahora es 4x2

print(mat_precios)


[[ 1.5  1. ]
 [ 2.   2.5]
 [ 5.   4.5]
 [16.  17. ]]


In [ ]:
# Matriz coste: cuánto paga cada persona en cada cafetería
mat_coste = np.dot(mat_bollos,mat_precios)

print(mat_coste)


[[50.  49. ]
 [58.5 61. ]
 [43.5 43.5]]


In [ ]:
cafeterias = ["Manolo Bakes", "Starbucks"]
clientes = ["Rafa", "Nacho", "Clara"]

for indice_cliente, nombre_cliente in enumerate(clientes):
    # Comparamos el coste en cada cafetería para ese cliente
    if mat_coste[indice_cliente][0] != mat_coste[indice_cliente][1]:
        idx_mejor_opcion = np.argmin(mat_coste[indice_cliente])
        mejor_local = cafeterias[idx_mejor_opcion]
        print(f"Para {nombre_cliente} la opción más barata es {mejor_local}")
    else:
        print(f"Para {nombre_cliente} el coste es el mismo en ambas cafeterías")

        

Para Rafa la opción más barata es Starbucks
Para Nacho la opción más barata es Manolo Bakes
Para Clara el coste es el mismo en ambas cafeterías


**BONUS (Voluntario):**  
Expresa utilizando markdown y LaTex las matrices y las operaciones que has realizado. Recuerda que para invocar LaTex denro de markdown se emplea $, y $$ si queremos que esté centrado. Y que [aquí](https://ashki23.github.io/markdown-latex.html#latex "Latex y Markdown") tienes muchos ejemplos de como trabajar con latex

$$\begin{bmatrix}6&5&3&1\\3&6&2&2\\3&4&3&1\end{bmatrix} \times \begin{bmatrix}1.5&1\\2&2.5\\5&4.5\\16&17\end{bmatrix} = \begin{bmatrix}50&49\\58.5&61\\43.5&43.5\end{bmatrix}$$

![image.png](attachment:image.png)

## PARTE II: Clasificador de Películas

II.1: Similitud entre películas

Vamos a calcular similitudes entre películas. Para ello las clasificasmos según generos, en concreto tenemos:

* Toy Story: Aventura, Animación, Comedia
* Jumanji: Aventura   
* Mortal Kombat: Acción, Aventura  
* Pocahontas: Animación, Romance

Y ahora calcularemos su similitud por el método del coseno. Para ello:

1. Convierte cada película en un vector y almacenalo en un diccionario con clave el nombre de la película. Pista: Para convertirlo en vector, crea un vector/lista por película con esta estructura (¿qué valores puedes darle por género?):

$$[Acción,  Aventura , Animación, Comedia, Romance]$$


In [ ]:
catalogo_peliculas = {
    "Toy Story":      [0, 1, 1, 1, 0],
    "Jumanji":        [0, 1, 0, 0, 0],
    "Mortal Kombat":  [1, 1, 0, 0, 0],
    "Pocahontas":     [0, 0, 1, 0, 1]
}




2. Crea una matriz de películas con un array de numpy. Muestra la dimension, shape, tamaño y tipo del array.


In [ ]:
vectores_valoracion = list(catalogo_peliculas.values())
matriz_valoraciones = np.array(vectores_valoracion)

print(matriz_valoraciones)


[[0 1 1 1 0]
 [0 1 0 0 0]
 [1 1 0 0 0]
 [0 0 1 0 1]]


In [ ]:
# Recomendación de película similar para cada una (sin Wall-E aún)
for pelicula_ref, vector_ref in catalogo_peliculas.items():
    mejor_similar = None
    sim_max = -1.0

    for pelicula_cand, vector_cand in catalogo_peliculas.items():
        if pelicula_ref == pelicula_cand:
            continue

        numerador = np.dot(vector_ref, vector_cand)
        denom = np.linalg.norm(vector_ref) * np.linalg.norm(vector_cand)
        coseno_sim = numerador / denom

        if coseno_sim > sim_max:
            sim_max = coseno_sim
            mejor_similar = pelicula_cand

    print(f"Para {pelicula_ref} la más similar es {mejor_similar}, similitud: {sim_max:.4f}")


Para Toy Story la más similar es Jumanji, similitud: 0.5774
Para Jumanji la más similar es Mortal Kombat, similitud: 0.7071
Para Mortal Kombat la más similar es Jumanji, similitud: 0.7071
Para Pocahontas la más similar es Toy Story, similitud: 0.4082



3. A partir de la similitud por coseno (recuerda a mayor coseno entre vectore más similitud), genera un programa que calcule la similitud entre cada pelicula y nos diga cuales son las más parecidas entre sí.


**BONUS (Voluntario)**:

A Jaime le encanta la premoritoria pelicula Wall-E (Acción, Aventura, Animación, Comedia), ¿qué película de las anteriores le recomendaríamos?

In [ ]:
# Añadimos una nueva película (Wall-E) y repetimos el análisis
catalogo_peliculas["Wall-E"] = [1, 1, 1, 1, 0]

for pelicula_ref, vector_ref in catalogo_peliculas.items():
    mejor_similar = None
    sim_max = -1.0

    for pelicula_cand, vector_cand in catalogo_peliculas.items():
        if pelicula_ref == pelicula_cand:
            continue

        numerador = np.dot(vector_ref, vector_cand)
        denom = np.linalg.norm(vector_ref) * np.linalg.norm(vector_cand)
        coseno_sim = numerador / denom

        if coseno_sim > sim_max:
            sim_max = coseno_sim
            mejor_similar = pelicula_cand

    print(f"Para {pelicula_ref} la más similar es {mejor_similar}, similitud: {sim_max:.4f}")


Para Toy Story la más similar es Wall-E, similitud: 0.8660
Para Jumanji la más similar es Mortal Kombat, similitud: 0.7071
Para Mortal Kombat la más similar es Jumanji, similitud: 0.7071
Para Pocahontas la más similar es Toy Story, similitud: 0.4082
Para Wall-E la más similar es Toy Story, similitud: 0.8660
